src/app/backend/resource/deployment/rollout.go

```diff
@@ -0,0 +1,111 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deployment
+
+import (
+	"github.com/kubernetes/dashboard/src/app/backend/api"
+	v1 "k8s.io/api/apps/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+
+	"errors"
+)
+
+// RollBackSpec is a specification for deployment rollback from an specific revision number
+type RollBackSpec struct {
+	// revision is the revision number of the replicateSet which we want to rollback
+	Revision string `json:"revision"`
+}
+
+// RollbackDeployment rollback to a specific replicaSet version
+func RollbackDeployment(client client.Interface, rollbackSpec *RollBackSpec, namespace, name string) error {
+	deployment, err := client.AppsV1().Deployments(namespace).Get(name, metaV1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	currRevision := deployment.Annotations["deployment.kubernetes.io/revision"]
+	if currRevision == "1" {
+		return errors.New("No revision for rolling back ")
+	}
+	matchRS, err := GetReplicateSetFromDeployment(client, namespace, name)
+	if err != nil {
+		return err
+	}
+	for _, rs := range matchRS {
+		if rs.Annotations["deployment.kubernetes.io/revision"] == rollbackSpec.Revision {
+			updateDeployment := deployment.DeepCopy()
+			updateDeployment.Spec.Template.Spec = rs.Spec.Template.Spec
+			_, err = client.AppsV1().Deployments(namespace).Update(updateDeployment)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+	return errors.New("No match revisionNumber replicateSet for deployment ")
+}
+
+// PauseDeployment is used to pause a deployment
+func PauseDeployment(client client.Interface, namespace, deploymentName string) (*v1.Deployment, error) {
+	deployment, err := client.AppsV1().Deployments(namespace).Get(deploymentName, metaV1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if deployment.Spec.Paused {
+		deployment.Spec.Paused = false
+		_, err = client.AppsV1().Deployments(namespace).Update(deployment)
+		if err != nil {
+			return nil, err
+		}
+		return deployment, nil
+	}
+	return nil, errors.New("the deployment is already paused")
+}
+
+// ResumeDeployment is used to resume a deployment
+func ResumeDeployment(client client.Interface, namespace, deploymentName string) (*v1.Deployment, error) {
+	deployment, err := client.AppsV1().Deployments(namespace).Get(deploymentName, metaV1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if !deployment.Spec.Paused {
+		deployment.Spec.Paused = true
+		_, err = client.AppsV1().Deployments(namespace).Update(deployment)
+		if err != nil {
+			return nil, err
+		}
+		return deployment, nil
+	}
+	return nil, errors.New("the deployment is already resumed")
+}
+
+// GetReplicateSetFromDeployment return all replicateSet which is belong to the deployment
+func GetReplicateSetFromDeployment(client client.Interface, namespace, deploymentName string) ([]v1.ReplicaSet, error) {
```

Comment:
s/GetReplicaSetFromDeployment/GetReplicaSetsFromDeployment

Do we really need to get all replica sets in the cluster? Replica sets inherit labels from deployment so we should be able to target only the ones created by specific deployment.

Reply:
oh, yes. I will delete this function, because I think it is the same as `handleGetDeploymentOldReplicaSets`

Reply:
It's not the same. `OldReplicaSets` are only returned in the middle of the rollout process. Most of the time list is empty.

Reply:
Oh, I wonder if it would be better for us to show all the historical versions in the deployment page?

Reply:
That is not `OldReplicaSets`. It would be a different feature.

Reply:
okay，I modified this function later. Thank you~

Reply:
Also not `Replicate` but `Replica`.

Reply:
Great~